homeassistant/components/vilfo/config_flow.py

```diff
@@ -0,0 +1,148 @@
+    except ValueError:
+        disallowed = re.compile(r"[^a-zA-Z\d\-]")
+        return all(x and not disallowed.search(x) for x in host.split("."))
+
+
+def _try_connect_and_fetch_basic_info(host, token):
+    """Attempt to connect and call the ping endpoint and, if successful, fetch basic information."""
+
+    # Perform the ping. This doesn't validate authentication.
+    controller = VilfoClient(host=host, token=token)
+    result = {"type": None, "data": {}}
+
+    try:
+        controller.ping()
+    except VilfoException:
+        result["type"] = RESULT_CANNOT_CONNECT
+        result["data"] = CannotConnect
+        return result
+
+    # Perform a call that requires authentication.
+    try:
+        controller.get_board_information()
+    except VilfoAuthenticationException:
+        result["type"] = RESULT_INVALID_AUTH
+        result["data"] = InvalidAuth
+        return result
+
+    if controller.mac:
+        result["data"][CONF_ID] = controller.mac
+        result["data"][CONF_MAC] = controller.mac
+    else:
+        result["data"][CONF_ID] = host
+        result["data"][CONF_MAC] = None
+
+    result["type"] = RESULT_SUCCESS
+
+    return result
+
+
+async def validate_input(hass: core.HomeAssistant, data):
+    """Validate the user input allows us to connect.
+
+    Data has the keys from DATA_SCHEMA with values provided by the user.
+    """
+
+    # Validate the host before doing anything else.
+    if not host_valid(data[CONF_HOST]):
+        raise InvalidHost
+
+    config = {}
+
+    result = await hass.async_add_executor_job(
+        _try_connect_and_fetch_basic_info, data[CONF_HOST], data[CONF_ACCESS_TOKEN]
+    )
+
+    if result["type"] != RESULT_SUCCESS:
+        raise result["data"]
+
+    # Return some info we want to store in the config entry.
+    result_data = result["data"]
+    config["title"] = f"{data[CONF_HOST]}"
+    config[CONF_MAC] = result_data[CONF_MAC]
+    config[CONF_HOST] = data[CONF_HOST]
+    config[CONF_ID] = result_data[CONF_ID]
+
+    return config
+
+
+class DomainConfigFlow(config_entries.ConfigFlow, domain=DOMAIN):
+    """Handle a config flow for Vilfo Router."""
+
+    VERSION = 1
+    CONNECTION_CLASS = config_entries.CONN_CLASS_LOCAL_POLL
+
+    async def async_step_user(self, user_input=None):
+        """Handle the initial step."""
+        errors = {}
+        if user_input is not None:
+            try:
+                info = await validate_input(self.hass, user_input)
+                await self.async_set_unique_id(info[CONF_ID])
+                self._abort_if_unique_id_configured()
+
+                return self.async_create_entry(title=info["title"], data=user_input)
+            except InvalidHost:
+                errors[CONF_HOST] = "wrong_host"
+            except CannotConnect:
+                errors["base"] = "cannot_connect"
+            except InvalidAuth:
+                errors["base"] = "invalid_auth"
+            except data_entry_flow.AbortFlow as abort_err:
```

Comment:
Why catch this and just reraise? Looks like we can remove this.

Reply:
This is to prevent it from being caught by the "unknown error"-clause, which otherwise catches it.

Reply:
Would a bare `raise` instead of `raise abort_err` be preferable?

Reply:
We should move setting unique_id etc out of this try...except. The try...except block should only contain the line that we except to raise.

Reply:
`else:` block added with the `unique_id` stuff in it. More readable and cleaner now :+1: